### ISSUE TITLE
Non-blocking consumer

### ISSUE THREAD

**Issue by aNutForAJarOfTuna** [Jroland/kafka-net#15]
Hi Jroland, 

The consumer seems only has a blocking consume function, do you think it's a good idea to add a non-blocking function allow us to specify an arbitrary number of messages to consume ? 

Thanks.

**Comment by buybackoff** [comment]
This is essential feature and I am surprised it was not there from the beginning. For even hundreds of topics a blocking consumer becomes unusable if each one blocks a real thread, but thousands of topics are not a rare use case. `IObservable` or `async Task<>` doesn't matter as one could be built from another, but a `Task` is simpler, e.g. a method `public async Task<Message> ConsumeNext(CancellationToken? cancellationToken = null)`.

I am looking into the code but not sure if I understand it enough. Is the `BlockingCollection` the only place where blocking occurs? What if we change `BlockingCollection` to [`AsyncColelction`](https://github.com/StephenCleary/AsyncEx/wiki/AsyncCollection) from the library [AsyncEx](https://github.com/StephenCleary/AsyncEx/wiki/AsyncCollection) that is already mentioned in #54 (or use `AsyncProducerConsumerQueue<T>` as described here http://stackoverflow.com/a/21225922/801189). Should this just work, or some deeper changes are required?

**Comment by Jroland** [comment]
The consumer needs a lot of work.  The current one in the library is not much modified from when I originally ported it over from python version.  It definitely needs some love.  Also the primary group creating kafka is working on a brand new design for the consumer as we speak...

https://cwiki.apache.org/confluence/display/KAFKA/Kafka+0.9+Consumer+Rewrite+Design

Because of this I have been mainly focused on the producer side of things.  The current consumer is functional, but by no means capable of supporting 100s of topics.  Definitely have a look at the consumer design linked above to get a better idea of direction.  

Also I think we can modify the current consumer pretty easily to make it a poll based consumer vs, the blocking one, or even a subscription based one similar to how EasynetQ does it.  This would help cover more topics at one time until we can get parity with Kafka 0.9.

**Comment by Jroland** [comment]
I like the IObservable pattern, and I have used it in the past.  It works well, and yah, I do think it would fit for this.  What I want to avoid though is pulling in more extenal dependencies into the base project.  (We would need to bring in the Reactive package for this wouldn't we?)  I like the way nancy seems to handle this, where you bring in the primary nancy package, and then add features by simply bringing in other packages.  This does require us to change the package to have a better dependancy inject design.

**Comment by buybackoff** [comment]
IObservable is push-based. It is possible to create it manually without Rx package, but is also possible to do so from a Task<Message> in a separate library. A simple Task-based async method in the minimum building block. With it one could use [lazy streams](http://hopac.github.io/Hopac/Hopac.html#def:module Hopac.Stream) from Hopac instead of Rx and keep everything pull-based.

Regarding the pull-request, is `BlockingCollection` the only thing that is blocking now? If so, I could add another method like 'public IObservable<Message> Observe()' without Rx dependency. Waiting for 0.9 could be long, but in the meantime these simplest methods could be enough. Even with underlying implementation changed, the method signatures will remain the same.

**Comment by Jroland** [comment]
I agree I don't want to wait for 0.9, we can definitely make the consumer better in the mean time.  The blocking collection is the only thing blocking the driver thread right now, so I would love to see what you put together.  I'll have a look at the pull, looks good at first glance.  Go ahead and add anything else you want to it.  Lets see what we can come up with for a better consumer.

**Comment by mikebridge** [comment]
This looks like a clean way of using Rx with Kafka in Scala:

   https://github.com/cjdev/kafka-rx
